File: LinkedList_problems/DesignBrowserHistory.js

// Problem link - https://leetcode.com/problems/design-browser-history/

/**
 * @param {string} homepage
 */
let BrowserHistory = function(homepage) {
    // Initialize the current page with the homepage
    this.page = {
        url: homepage, // URL of currnet page
        back: null,    // reference to the previous page
        next: null,    // reference to the next page
    };
};

/** 
 * @param {string} url
 * @return {void}
 */
BrowserHistory.prototype.visit = function(url) {
    // Create a new page object for the visited URL
    this.page.next = {
        url,             // set URL of the new page
        back: this.page, // set URL of the previous page
        next: null          // there is no next page yet
    };
    // update current page to the new page
    this.page = this.page.next;
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.back = function(steps) {
    // Traverse back through the linked list of pages
    // reaching the beginning of the history or the specified number of steps
    while(this.page.back && steps){
        // Move to the previous page
        this.page = this.page.back;
        // Decrement the steps counter
        steps--
    }

    // return current page URL
    return this.page.url;
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.forward = function(steps) {
    // Traverse forward through the linked list of pages
    // reaching the end of the history or the specified number of steps
    while(this.page.next && steps){
        // Move to the next page
        this.page = this.page.next;
        // Decrement the steps counter
        steps--;
    }
    // return current page URL
    return this.page.url;
};

// test above code
// Create a new browser history with a homepage
let history = new BrowserHistory("leetcode.com");

// Visit some pages
history.visit("google.com");
history.visit("facebook.com");
history.visit("youtube.com");

// Navigate back two steps
console.log(history.back(2)); // Output: google.com
console.log(history.back(1))
console.log(history.forward(2))
console.log(history.back(3))
console.log(history.forward(7))

// Visit another page
history.visit("hackerrank.com");

File: 06.LinkedList_problems/DesignBrowserHistory.cpp

// Problem link - https://leetcode.com/problems/design-browser-bw/

#include<bits/stdc++.h>
using namespace std;

class BrowserHistory{
private:
    class Node{
    public:
        string url;
        Node* back;
        Node* next;

        Node(string url){
            this->url = url;
            this->back = nullptr;
            this->next = nullptr;
        }
    };

    Node* page;

public:
    BrowserHistory(string homepage){
        this->page = new Node(homepage);
    }

    void visit(string url){
        this->page->next = new Node(url);
        this->page->next->back = this->page;
        this->page = this->page->next;
    }

    string back(int steps){
        while(steps && this->page->back != nullptr){
            this->page = this->page->back;
            steps--;
        }
        return this->page->url;
    }

    string forward(int steps){
        while(steps && this->page->next != nullptr){
            this->page = this->page->next;
            steps--;
        }
        return this->page->url;
    }
};

int main(){
    BrowserHistory history("leetcode.com");
    history.visit("google.com");
    history.visit("facebook.com");
    history.visit("youtube.com");

    cout << history.back(2) << endl; // Output: google.com
    cout << history.back(1) << endl; // Output: leetcode.com
    cout << history.forward(2) << endl; // Output: facebook.com
    cout << history.back(3) << endl; // Output: leetcode.com
    cout << history.forward(7) << endl; // Output: youtube.com

    history.visit("hackerrank.com");
    return 0;
}

